src/llvm-remove-addrspaces.cpp

```diff
@@ -0,0 +1,461 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#include "llvm-version.h"
+
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Verifier.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/InstIterator.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/Support/Debug.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+#include <llvm/Transforms/Utils/ValueMapper.h>
+
+#include "codegen_shared.h"
+#include "julia.h"
+
+#define DEBUG_TYPE "remove_addrspaces"
+
+using namespace llvm;
+
+using AddrspaceRemapFunction = std::function<unsigned(unsigned)>;
+
+
+//
+// Helpers
+//
+
+class AddrspaceRemoveTypeRemapper : public ValueMapTypeRemapper {
+    AddrspaceRemapFunction ASRemapper;
+
+public:
+    AddrspaceRemoveTypeRemapper(AddrspaceRemapFunction ASRemapper)
+        : ASRemapper(ASRemapper)
+    {
+    }
+
+    Type *remapType(Type *SrcTy)
```

Comment:
This algorithm here doesn't appear to have a guaranteed fixed point (it doesn't deal with the fact that `PointerType->getElementType` can form cycles in the type tree). We avoid emitting any code like that, so it's probably okay for us, but probably not for upstream.

Reply:
How does LLVM typically deal with that? A `Seen` set?

Reply:
from @vtjnash: https://llvm.org/doxygen/IRMover_8cpp_source.html#l00235